### ISSUE TITLE
[rllib] What is the proper way to restore checkpoint for fine-tuning / rendering / evaluation of a trained agent based on example/multiagent_cartpole.py?

### ISSUE THREAD

**Issue by pengzhenghao** [ray-project/ray#4569]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Ray installed from (source or binary)**: pip install ray
- **Ray version**: 0.6.5
- **Python version**: 3.6.2
- **Exact command to reproduce**:

### Describe the problem

Before my question, let me introduce my understanding of the checkpoint file system. (you can skip it and toward my question)

The codes in `example/multiagent_cartpole.py` produces a `experiment_state-2019-04-03_00-47-28.json`-like file and a directory `PPO_experiment_name` with a few .pkl, .json, .csv files in it.

 The file system looks like:

```
- local_dir (say: "~/ray_results")
    - exp_name (say: "PPO")
        - checkpoints (say: experiment_state-2019-04-05_17-59-00.json)
        - directory (named like: PPO_cartpole_0_2019-04-05_18-28-0296h2tknq)
            - xxx.log
            - params.json
            - params.pkl (This is the file to store trained parameter, I guess?)
            - progress.csv
            - result.json
```

After one successful training, now we have a trained agent (Because I used one shared policy for all agent). We set the `local_dir` exactly the same as training. Then set the `exp_name` exactly as training too, namely `PPO`.

Now it's my problem. The `tune.run` function take two arguments which looks like helpful for restoring.

## "resume" argument

The resume argument, once set to True, will automatically search in `local_dir/exp_name/` finding the most recent `experiment_state-<date_time>.json`. 

 The `resume` work well. After setting it to true, the restoring seems to be successful, but the program immediately terminated, as if it inherit the termination states from the checkpoint.

Here's the log:

```
== Status ==
Using FIFO scheduling algorithm.
Resources requested: 0/12 CPUs, 0/1 GPUs
Memory usage on this node: 4.3/16.7 GB
Result logdir: /home/SENSETIME/pengzhenghao/ray_results/PPO
Number of trials: 1 ({'TERMINATED': 1})
TERMINATED trials:
 - PPO_tollgate_0:	TERMINATED, [12 CPUs, 1 GPUs], [pid=9214], 4846 s, 300 iter, 1320000 ts, 1.1e+03 rew
```

The printed reward is exactly what trained agent able to give, but I cannot continue to train this agent, even if I set the `num_iters` greater than the number of iterations in last training (namely 300).

What's more, it seems impossible using the `resume` argument to specify a checkpoint given the exact filename.

In a nut shell, my question on the `resume` argument is:

1. What's the meaning for this argument? It seems like it's only used for restore checkpoint from unexpected failures. Therefore, it cannot be used to restore specified checkpoint. Am I correct?

## "restore" argument

After setting `restore=<log_dir>`, namely `restore="./experiments"`, which is my `log_dir`, it turn out to be an error:

```
Traceback (most recent call last):
  File "xxx/anaconda3/envs/dev/lib/python3.6/site-packages/ray/tune/ray_trial_executor.py", line 499, in restore
    ray.get(trial.runner.restore.remote(value))
  File "xxx/anaconda3/envs/dev/lib/python3.6/site-packages/ray/worker.py", line 2316, in get
    raise value
ray.exceptions.RayTaskError: ray_PPOAgent:restore() (pid=28099, host=g114e1900387)
  File "xxx/anaconda3/envs/dev/lib/python3.6/site-packages/ray/tune/trainable.py", line 304, in restore
    with open(checkpoint_path + ".tune_metadata", "rb") as f:
FileNotFoundError: [Errno 2] No such file or directory: './experiments.tune_metadata'
```

I have checked everywhere of this computer and there is no such a file ended with `.tune_metadata`. I am really confusing.

In short, what I am trying to do is:

1. **Restore the trained agent and continue it's training with the same config.**

2. **Restore the trained agent, retrieve the Policy network, and used in the same environment with rendering, in order to visualize it's performance.**

3. **Restore the trained agent as a pre-trained agent and modify the config, such as using more workers and GPU to training on cluster.**

Could you please tell me what I should do?

(By the way, the document is really insufficient for thoroughly understanding the whole process of rllib. Nevertheless I still appreciate your guys for this excellent project, wish some day I can make some contribution too~)

**Comment by ericl** [comment]
I think there is some confusion here about tune's checkpointing of experiment state, vs RLlib's checkpointing of trial state.

To enable RLlib checkpointing, you have to specify `--checkpoint-freq`. For example: `rllib train --run=PG --checkpoint-freq=1 --env=CartPole-v0`

Then, this will create checkpoints in ~/ray_results that includes the .tune_metadata file. To restore, you can specify one of those paths, for example `rllib train --run=PG --env=CartPole-v0 --restore=$HOME/ray_results/default/PG_CartPole-v0_0_2019-04-05_16-43-02s_gcpmkl/checkpoint_9/checkpoint-9`.

This path call also be passed to agent.restore() in the Python API, which allows support for more advanced use cases like (2). For (1) and (3) I think the --restore flag for Tune may work.

> By the way, the document is really insufficient for thoroughly understanding the whole process of rllib. 

Agree! This part happens to be halfway documented in RLlib and half in Tune. Some of it is here: https://ray.readthedocs.io/en/latest/rllib-training.html Any suggestions on how to improve this would be helpful.

**Comment by PengZhenghao** [comment]
Thanks for your reply! I find that if running tune.run without any change, the parameter of trained agent would not saved... The .pkl file saved automatically simply records “trail” instead of anything related to neural network.

Unfortunately the training for last few days goes in vain. I suggest to turn checkpoint_final be True as default...

**Comment by ericl** [comment]
The only blocker for enabling checkpointing by default is https://github.com/ray-project/ray/pull/4490

That will avoid out of disk space errors for long training.

**Comment by PengZhenghao** [comment]
For the potential reader:

`resume` argument do nothing but continue the last unfinished task. In this mode, it's no allowed to reset the `num_iters`.

`restore` argument take the path of the checkpoint file as input. Concretely, the file look like `~/ray_results/expname/envname_date_someothercodes/checkpoint_10/checkpoint-10`. Note that the checkpoint files would only exist for those `tune.run()` executions with `checkpoint_at_end=True` or `checkpoint_freq` setting to non-zero value.

Using `restore` argument and taking the checkpoint from which you want to continue the experiment is the only way to enlarge the number of iterations of a finished or unfinished experiment.

Thank Eric for offering quick and kind responses!

**Comment by evanatyourservice** [comment]
For me, the restore does not work no matter what I try it seems.

    tune.run(
    "PPO",
    #name="PPO_discrete5",
    local_dir="/content/drive/My Drive/Colab Notebooks/rltrader/Experiments",

    checkpoint_freq=10, # iterations
    checkpoint_at_end=True,
    max_failures=100,
    #resume=True,
    restore='content/drive/My Drive/Colab Notebooks/rltrader/Experiments/PPO_discrete3/PPO_ContTradingEnv_0_2019-05-03_04-51-02zykryvgl/checkpoint_218/checkpoint-218',
    
    #search_alg=algo,
    #scheduler=ahb,
    # 2 if testing, 50 or more for real
    #num_samples=50,

    stop={
        # "episode_reward_mean": 0,
        # "training_iteration": 1,
        # "timesteps_total": 1000,
        "episodes_total": 1000,
    },

What combination of the above do I need to restore a checkpoint using tune.run, or is restore not working? I have run 1000 episodes, and wish to run 1000 more.

**Comment by pengzhenghao** [comment]
@evanatyourservice Are you using the latest Ray? The ray failed to restore due to a bug and still not fixed yet. See https://github.com/ray-project/ray/pull/4733

**Comment by pengzhenghao** [comment]
@evanatyourservice Please re-run your codes using the latest Ray and see if everything work well.

**Comment by lsisoft** [comment]
My view of the resume/restore:
- you start a training with a lot of grid search for a number of interations, let's say on some gamma options
- you add checkpoint_freq=10, checkpoint_at_end=True to ray.tune
- you put some unique name name=name and you specify local_dir in some folder you have space. Name paramter is the folder of this experiment (important)

If somehow the running stops, you add resume=True. If any of the trials give an error, resume won't restart it.

Here comes the nice part:
- if you get an errored trail, let's say due to memory, you add restore parameter. How to do that: run
<pre>find "[local_dir]/[name]" -iname checkpoint-[K]</pre> where K is the last checkpoint created, or last iteration 
- when you do that, you're out of grid search mode, you need to specify gamma, lr, number of iterations manually, but this is the good part, as it allows you to resume any past checkpoint and go longer number of iterations or try other things (gamma/lr can be changed on a checkpoint)

Other ideas:
**LR:**
- try a constant lr or an lr_schedule, like:
<pre>
        'lr_schedule': [[0 * lr_batch_size, 5e-5],
                        [75 * lr_batch_size, 5e-5],
                        [110 * lr_batch_size, 1e-5],
                        [110 * lr_batch_size, 1e-5],
                        [120 * lr_batch_size, 5e-6],
                        [140 * lr_batch_size, 5e-7],
                        [200 * lr_batch_size, 1e-10],
                        [300 * lr_batch_size, 1e-12],
                        ]
</pre> where lr_batch_size is your number of timesteps per interation.
- the learning rate can be resumed at any time from a checkpoint

**Chart:**
- plot episode_mean_reward by training_iteration, and also by training_interation you can chose to plot:
  - episode_reward_min, episode_reward_max
  - cur_lr, look for the column in the dataframe ending in "/cur_lr"

**Comment by stefanbschneider** [comment]
Since I was searching for a simple way to load a trained agent and continue training with RLlib, and I only found this issue, here's what I found & what's the easiest way in my opinion:

```python
ray.tune.run(PPOTrainer, config=myconfig, restore=path_to_trained_agent_checkpoint)
```

Ie, just set the path in the `restore` argument, that's it! No need for a custom train function.

**Comment by FruechteBini** [comment]
@stefanbschneider i am currently working on the same issue - what is the parameter you use in the config equivalent to your test() function (evaluation mode) in this issue: https://github.com/ray-project/ray/issues/9123#issuecomment-649543419 ?

**Comment by stefanbschneider** [comment]
@FruechteBini Not sure, I get your question correctly. In

```python
ray.tune.run(PPOTrainer, config=myconfig, restore=path_to_trained_agent_checkpoint)
```

from the example above, `myconfig` is the [RLlib config dict](https://docs.ray.io/en/latest/rllib-training.html#configuration).

In the issue comment you referenced, I did

```python
def test(self):
    """Test trained agent for a single episode. Return the episode reward"""
    # instantiate env class
    env = self.env_class(self.env_config)
```

inside my testing/evaluation function. Here, `self.env_config` is just the dict that I pass to instantiate my environment, i.e., env settings, but nothing about RLlib.

Does that answer your question?

**Comment by FruechteBini** [comment]
Hey @stefanbschneider ! Thanks for the answer and sorry for the late reply, just saw it. I think i figured it out - i was curious if there is a parameter that makes sure e.g. "explore" is set to False - but i guess the "restore"-Option does this. Thanks again, I will test this later.

**Comment by stefanbschneider** [comment]
@FruechteBini No worries. Yes, there is a parameter inside the RLlib config that you can set to disable exploration: `config['explore'] = explore`; it is not set automatically when you call restore!

**Comment by felipeeeantunes** [comment]
I'm trying to restore from an s3 bucket without setting `resume="REMOTE"` and getting `FileNotFoundError: [Errno Path does not exist]`. The solution described here seems to work with a local directory but not when using a remote one.

**Comment by quinn-dougherty** [comment]
Now comes the question of inferring latest checkpoints automatically. After reading this thread I concluded that tune doesn't have this built in, so I rolled this (based on [this](https://stackoverflow.com/questions/2731014/finding-most-recently-edited-file-in-python))

```
def all_dirs_under(path: str):
    """Iterates through all dirs that are under the given path."""
    for cur_path, dirnames, filenames in os.walk(path):
        for dir_ in dirnames:
            yield os.path.join(cur_path, dir_)

def latest_dir(path: str) -> str:
    return max(all_dirs_under(path), key=os.path.getmtime)

def walk_to_checkpoint(to_restore_path: str): 
    """Return None if checkpoint isn't there"""
    for cur_path, dirnames, _ in os.walk(to_restore_path):
        for dir_ in dirnames:
            if dir_.startswith("train_fn_"):
                # print(os.path.join(cur_path, dir_))
                for cur_path_, dirnames_, _ in os.walk(os.path.join(cur_path, dir_)):
                    for dir__ in dirnames_:
                        if dir__.startswith("checkpoint_"):
                            return os.path.join(cur_path_, dir__)
```

**Comment by stefanbschneider** [comment]
@quinn-dougherty I'm currently working on integrating a function `get_last_checkpoint()` to simplify this. Issue: https://github.com/ray-project/ray/issues/7927, PR: https://github.com/ray-project/ray/pull/14735

We're currently still debating what we want the function should do exactly? Load the last checkpoint in terms of creation time or in terms of number of training steps? If there are multiple trials in the directory (eg, belonging to different hyperparameters), which trial should be loaded (by default)? The last one or the best one (in terms of some metric)?

**Comment by borninfreedom** [comment]
> For me, the restore does not work no matter what I try it seems.
> 
> ```
> tune.run(
> "PPO",
> #name="PPO_discrete5",
> local_dir="/content/drive/My Drive/Colab Notebooks/rltrader/Experiments",
> 
> checkpoint_freq=10, # iterations
> checkpoint_at_end=True,
> max_failures=100,
> #resume=True,
> restore='content/drive/My Drive/Colab Notebooks/rltrader/Experiments/PPO_discrete3/PPO_ContTradingEnv_0_2019-05-03_04-51-02zykryvgl/checkpoint_218/checkpoint-218',
> 
> #search_alg=algo,
> #scheduler=ahb,
> # 2 if testing, 50 or more for real
> #num_samples=50,
> 
> stop={
>     # "episode_reward_mean": 0,
>     # "training_iteration": 1,
>     # "timesteps_total": 1000,
>     "episodes_total": 1000,
> },
> ```
> 
> What combination of the above do I need to restore a checkpoint using tune.run, or is restore not working? I have run 1000 episodes, and wish to run 1000 more.

What's the difference between "training_iteration" and "episodes_total" ?

**Comment by borninfreedom** [comment]
What's the difference between "training_iteration" and "episodes_total" ?

**Comment by stefanbschneider** [comment]
@borninfreedom I think a training iteration consists of collecting and training on experience from one "training batch", i.e., it collects data until the training batch is full, independent of whether this experience is from just one episode or multiple episodes.
An episode is environment-specific, e.g., terminates whenever the pole tips and falls in the common cart pole env.
Hence, collecting enough experience for one training iteration may require less than one episode (e.g., if episodes are long or even infinite), exactly one episode (by chance), or even multiple episodes.

**Comment by Roller44** [comment]
> @quinn-dougherty I'm currently working on integrating a function `get_last_checkpoint()` to simplify this. Issue: #7927, PR: #14735
> 
> We're currently still debating what we want the function should do exactly? Load the last checkpoint in terms of creation time or in terms of number of training steps? If there are multiple trials in the directory (eg, belonging to different hyperparameters), which trial should be loaded (by default)? The last one or the best one (in terms of some metric)?
> 
> **Update:** The PR is merged and the function is now available. Documented here: https://docs.ray.io/en/master/rllib-training.html#basic-python-api
> 
> ```python
> # or simply get the last checkpoint (with highest "training_iteration")
> last_checkpoint = analysis.get_last_checkpoint()
> # if there are multiple trials, select a specific trial or automatically
> # choose the best one according to a given metric
> last_checkpoint = analysis.get_last_checkpoint(
>     metric="episode_reward_mean", mode="max"
> )
> ```

If I want to restore a trained agent from the best trial, should I do the following?
`agent.restore(last_checkpoint)`

**Comment by stefanbschneider** [comment]
> If I want to restore a trained agent from the best trial, should I do the following?
agent.restore(last_checkpoint)

@Roller44 Yes, exactly!

**Comment by rfali** [comment]
Thanks a lot for this PR @stefanbschneider!

**Comment by stefanbschneider** [comment]
@rfali I'm happy to hear the PR is useful!

1. Yes, that's correct.
2. Yes, `ray.tune.run()` returns an ExperimentAnalysis object, on which you can call methods like `get_last_checkpoint()` (see [docs](https://docs.ray.io/en/latest/tune/api_docs/execution.html#tune-run)). So you could do:
```python
analysis = ray.tune.run(...)
checkpoint = analysis.get_last_checkpoint()
```
3. Yes, there are both methods, `get_last_checkpoint()` and `get_best_checkpoint()`. 
[`get_last_checkpoint(trial, metric, mode)`](https://docs.ray.io/en/latest/tune/api_docs/analysis.html#ray.tune.ExperimentAnalysis.get_last_checkpoint) loads the checkpoint with the highest training iteration, i.e., longest training. You can specify a specific trial as optional argument or it loads the checkpoint with highest training iteration from all trials.
[`get_best_checkpoint(trial, metric, mode)`](https://docs.ray.io/en/latest/tune/api_docs/analysis.html#ray.tune.ExperimentAnalysis.get_best_checkpoint) is more generic. It requires you passing a specific trial (i.e., `trial` must not be `None`) and you can configure any metric and mode you like, e.g., to load the checkpoint with highest mean reward. You can specify metric and mode either as arguments on each call or just set them once as object attributes (`self.default_metric` and `self.default_mode`). If you set neither, the default metric are training iterations and the method does the same as `get_last_checkpoint()` (except you have to specify a trial).

I hope this helps?

**Comment by richielo** [comment]
Hello @stefanbschneider, I am still a bit confused after reading through this issue and have been having trouble loading a trained model. I set my config as this:

```
   analysis = tune.run(
        "PPO", 
        # name="scheduling",
        name="exp_name",
        config=config,
        checkpoint_freq=500, #How frequently I want to save policies to the disk
	local_dir="./results",
        fail_fast="raise",
        stop={"timesteps_total": 100 * 10 ** 5}, #stopping criteria; I am using number of time steps, could use for instance reward achieved
        # resume = True
    )
```

And when I try to load one of the checkpoints in under "./results":

```
   agent = PPO.PPOTrainer(config=load_agent_config(), env="env_name")
    # Load model 
    agent.restore('results/exp_name/checkpoint_path/')
```

I am getting the error of missing '.tune_metadata'. Am I missing something here? I thought setting checkpoint_freq should already solve the problem. Thanks a lot in advance.

**Comment by dylan906** [comment]
Having read through this thread and the documentation on [Tuner](https://docs.ray.io/en/latest/tune/api_docs/execution.html#tune-run), I am still confused as to how to continue training from a checkpoint that had previously completed without error.

How do you restore a tuner from a specific trial checkpoint? `ray.tune.Tuner.restore()` seems to only work if you input an experiment directory. But if an experiment has multiple trials in it, you don't have control over which checkpoints are used to restore the tuner.

**Comment by MatFl** [comment]
I am having the same troubles finding a solution on how to restore a (specific) checkpoint when using the Tuner.
@dylan906 did you manage to find a solution?

**Comment by dylan906** [comment]
> I am having the same troubles finding a solution on how to restore a (specific) checkpoint when using the Tuner. @dylan906 did you manage to find a solution?

@MatFl , no I was not able to find a solution.

**Comment by MatFl** [comment]
I dug a bit deeper and found a workaround for this.
The Tuner init has a _tuner_kwargs dict which probably should not be used, but is the only working solution I found.
https://docs.ray.io/en/latest/tune/api/doc/ray.tune.Tuner.__init__.html#ray.tune.Tuner.__init__

These kwargs will be passed to tune.run which has a restore arg that works. So passing a dict with the 'restore' key and the checkpoint path to the _tuner_kwargs works for me.
